compress/src/lib.rs

```diff
@@ -0,0 +1,34 @@
+/// Compresses the passed string into the output byte vector.
+pub fn compress(input: String) -> Vec<u8> {
+    let input = input.into_bytes();
+    let mut output = Vec::with_capacity(input.len());
+    brotli::BrotliCompress(&mut input.as_slice(), &mut output, &Default::default()).unwrap();
+    output
+}
+
+#[macro_export]
+macro_rules! aoc_input {
+    ($path:literal) => {{
+        let file = include_bytes!("../input.compressed").to_vec();
+        compress::decompress(file)
+    }};
+}
+
+/// Decompresses the passed byte vector into a string.
+pub fn decompress(input: Vec<u8>) -> String {
+    let mut output = Vec::with_capacity(input.len());
+    brotli::BrotliDecompress(&mut input.as_slice(), &mut output).unwrap();
```

Comment:
So there's a lot to say here. I'll leave high level overviews and we can dig into anything else if we feel like it!
1. The reason why we need a `&mut` here is not because of anything to do with the `Read` trait - it's just required by the [`BrotliDecompress` API](https://docs.rs/brotli/latest/brotli/fn.BrotliDecompress.html).
2. According to [the Rust API guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html?highlight=read#generic-readerwriter-functions-take-r-read-and-w-write-by-value-c-rw-value), `BrotliDecompress` really should accept readers and writers by value, not by `&mut`. However making the switch is a breaking change! First, to see why this is surprising, look at [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f3ec5e2c9bbf9889b20181b115520fa0) showing it "isn't" a breaking change. Next, look at [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5d08dbe72f0aa40adf9b83080245f8aa) showing it _is_ a breaking change! What's going on here?? It turns out to be a fairly complicated topic called [`reborrowing`](https://github.com/rust-lang/reference/issues/788). It doesn't [work trivially with generics](https://stackoverflow.com/a/48144226) so there's a bit of tension here between the Rust API guidelines recommending something for flexibility while that flexibility also requires [surprise manual reborrowing](https://github.com/launchbadge/sqlx/issues/257). Either way, we can't just make a PR to update it :-)

Reply:
Oh and another thing that was originally weird - we wondered why the `std::io::Read` trait took `&mut self` all the time since reading sounded like something you could do in a shared setting. I imagine this is because reading from a network socket or file requires updating internal state that's unsafe to do with a shared reference.